pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -225,3 +252,26 @@ func (p *ecrProvider) Provide() credentialprovider.DockerConfig {
 	}
 	return cfg
 }
+
+// getCurrentRegion queries the local EC2 metadata service and returns the name
+// of the region in which the current host is running
+func getCurrentRegion() (string, error) {
+	// Without a timeout this hangs forever. Use a really agressively low
+	// timeout because the ec2metadata service is bound to a local interface on
+	// the EC2 hosts.
+	sess, err := session.NewSession(aws.NewConfig().
+		WithHTTPClient(&http.Client{Timeout: 3 * time.Millisecond}).
```

Comment:
3 millisecond does still seem _very_ aggressive, but I guess we're doing this because we don't know if we are on AWS?

cadvisor does this https://github.com/google/cadvisor/blob/master/utils/cloudinfo/aws.go#L27-L38 - could we do the same?

Reply:
Correct, this timeout exists so we don't hang on module init if we're not running on an AWS instance. This is querying the local instance metadata service (bound to `169.254.169.254`) so there's not an actual network hop involved and in practice 1ms worked suitably well.

I don't think the cadvisor approach reliably works. I checked on a test instance and the "file" they use is empty. On newer systems [/etc/os-release](https://www.freedesktop.org/software/systemd/man/os-release.html) is the right thing to use but older versions of Amazon Linux don't ship that and rely instead on `/etc/system-release`.

I don't know of a good reliable way to validate that a host is running on AWS across all OS distributions and versions other than querying the metadata service and hoping it responds. I'm happy to raise this timeout if you think it's too low or swap to a different detection approach if you know a better one.

Reply:
@kubernetes/sig-aws-misc for suggestions and eyes on whether this approach makes sense and whether 3ms is as aggressive as it seems to me